src/core/recv_buffer.h

```diff
@@ -122,12 +140,24 @@ QuicRecvBufferHasUnreadData(
 // Changes the buffer's virtual buffer length.
 //
 _IRQL_requires_max_(DISPATCH_LEVEL)
-void
-QuicRecvBufferIncreaseVirtualBufferLength(
+BOOLEAN
+QuicRecvBufferTryIncreaseVirtualBufferLength(
     _In_ QUIC_RECV_BUFFER* RecvBuffer,
     _In_ uint32_t NewLength
     );
 
+//
+// Provide externally allocated chunks to the buffer.
+// At least one chunk must be provided.
+// Only valid for QUIC_RECV_BUF_MODE_EXTERNAL mode.
+//
+_IRQL_requires_max_(DISPATCH_LEVEL)
+QUIC_STATUS
+QuicRecvBufferProvideChunks(
+    _Inout_ QUIC_RECV_BUFFER* RecvBuffer,
+    _Inout_ CXPLAT_LIST_ENTRY* /* QUIC_RECV_CHUNKS */ Chunks
```

Comment:
This assumes the caller allocates space and initializes a `QUIC_RECV_CHUNK` struct, instead of an array of `QUIC_BUFFER`s? While that simplifies this interface, I'm not sure that is reasonable.

Reply:
The idea is to expose an API that takes an array of QUIC_BUFFERs as a parameter, convert it inline to a list of QUIC_RECV_CHUNKS (which takes memory allocation and could fail), then call `QuicRecvBufferProvideChunks` in the workqueue.
This way, we can keep `QuicRecvBufferProvideChunks` mostly failure free (the overflow is still possible, but it would be reasonable to kill the connection in that case I think).

I was trying to keep the API and recvbuffer parts separate to simplify the review, but it might be starting to backfire. I might push everything at once so you have the full context.